// @icedesign/form-binder/lib/FormError.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(source, true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var FormError =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2.default)(FormError, _Component);

  function FormError() {
    (0, _classCallCheck2.default)(this, FormError);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(FormError).apply(this, arguments));
  }

  (0, _createClass2.default)(FormError, [{
    key: "render",
    value: function render() {
      var path = this.props.name;
      var errors = this.context.getError(path);
      if (errors.length === 0) return null;
      var render = this.props.render;

      if (render && typeof render === 'function') {
        return render(errors);
      }

      var _this$props = this.props,
          className = _this$props.className,
          style = _this$props.style;
      return _react.default.createElement("span", {
        className: className,
        style: _objectSpread({
          fontSize: 12,
          color: 'red'
        }, style)
      }, errors.map(function (error) {
        return error.message;
      }).join('，'));
    }
  }]);
  return FormError;
}(_react.Component);

exports.default = FormError;
FormError.displayName = 'FormError';
FormError.propTypes = {
  /**
   * 自定义渲染报错的组件和处理逻辑
   */
  render: _propTypes.default.func,
  className: _propTypes.default.string,
  style: _propTypes.default.object
};
FormError.defaultProps = {
  render: undefined,
  className: '',
  style: {}
};
FormError.contextTypes = {
  getError: _propTypes.default.func
};
