// @jxkang/web-cmpt/lib/preview-image/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _slicedToArray = function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; }(); /**
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          * @Author: 谭生虎 TanShenghu tanshenghu@163.com
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          * @Update: 2019-12-15
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          * @Description: 图片预览组件
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          */


var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _icon = require('../icon');

var _icon2 = _interopRequireDefault(_icon);

var _portalDom = require('../portal-dom');

var _portalDom2 = _interopRequireDefault(_portalDom);

var _style = require('./style');

var _style2 = _interopRequireDefault(_style);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var noop = function noop() {};
var g = window;
var Body = g.document.body || g.document.getElementsByTagName('body')[0];

var previewImage = function previewImage(_ref) {
  var _ref$images = _ref.images,
      images = _ref$images === undefined ? [] : _ref$images,
      _ref$preImgWidth = _ref.preImgWidth,
      preImgWidth = _ref$preImgWidth === undefined ? 800 : _ref$preImgWidth,
      className = _ref.className,
      _ref$scope = _ref.scope,
      scope = _ref$scope === undefined ? noop : _ref$scope,
      _ref$onChangeUp = _ref.onChangeUp,
      onChangeUp = _ref$onChangeUp === undefined ? noop : _ref$onChangeUp,
      _ref$onChangeDown = _ref.onChangeDown,
      onChangeDown = _ref$onChangeDown === undefined ? noop : _ref$onChangeDown,
      _ref$onRemove = _ref.onRemove,
      onRemove = _ref$onRemove === undefined ? noop : _ref$onRemove;

  var _useState = (0, _react.useState)(false),
      _useState2 = _slicedToArray(_useState, 2),
      zoom = _useState2[0],
      setZoom = _useState2[1];

  var _useState3 = (0, _react.useState)(false),
      _useState4 = _slicedToArray(_useState3, 2),
      visiblePreview = _useState4[0],
      setVisiblePreview = _useState4[1];

  /**
   * 向上移动图片
   */


  var handleSetUp = function handleSetUp() {
    if ((arguments.length <= 0 ? undefined : arguments[0]) > 0) {
      onChangeUp.apply(undefined, arguments);
    }
  };

  /**
   * 移除图片
   */
  var handleRemove = function handleRemove() {
    onRemove.apply(undefined, arguments);
  };

  /**
   * 向下移动图片
   */
  var handleSetDown = function handleSetDown() {
    if ((arguments.length <= 0 ? undefined : arguments[0]) < images.length - 1) {
      onChangeDown.apply(undefined, arguments);
    }
  };

  /**
   * 图片全屏预览缩放
   */
  var handleZoom = function handleZoom() {
    setZoom(!zoom);
  };

  /**
   * 调用 展开全屏预览
   */
  var showPreview = function showPreview() {
    setVisiblePreview(true);
    Body.style = (Body.style.cssText || '') + 'overflow:hidden';
  };

  /**
   * 调用 关闭全屏预览
   */
  var closePreview = function closePreview() {
    setVisiblePreview(false);
    Body.style.cssText = Body.style.cssText.replace(/overflow:\s*hidden;?/, '');
  };

  scope({
    showPreview: showPreview
  });

  return _react2.default.createElement(
    'section',
    { className: (0, _classnames2.default)(_style2.default.images_wraper, className) },
    _react2.default.createElement(
      'ul',
      null,
      Array.isArray(images) && images.map(function (item, idx) {
        return _react2.default.createElement(
          'li',
          { key: idx, className: _style2.default.items },
          _react2.default.createElement(
            'figure',
            null,
            _react2.default.createElement('img', { src: item, alt: item })
          ),
          _react2.default.createElement(
            'div',
            { className: _style2.default.handler_box },
            _react2.default.createElement(
              'span',
              { className: idx === 0 ? _style2.default.disabled : null, onClick: function onClick() {
                  return handleSetUp(idx, item);
                } },
              _react2.default.createElement(_icon2.default, { type: 'jiantou' })
            ),
            _react2.default.createElement(
              'span',
              { onClick: function onClick() {
                  return handleRemove(idx, item);
                } },
              '\xD7'
            ),
            _react2.default.createElement(
              'span',
              { className: idx === images.length - 1 ? _style2.default.disabled : null, onClick: function onClick() {
                  return handleSetDown(idx, item);
                } },
              _react2.default.createElement(_icon2.default, { type: 'jiantou' })
            )
          )
        );
      })
    ),
    _react2.default.createElement(
      _portalDom2.default,
      null,
      visiblePreview ? _react2.default.createElement(
        'section',
        { className: _style2.default.preview_dialog },
        _react2.default.createElement(
          'div',
          { className: _style2.default.preview_close, title: '\u5173\u95ED', onClick: closePreview },
          '\xD7'
        ),
        _react2.default.createElement(
          'ol',
          {
            className: (0, _classnames2.default)(_style2.default.preview_ol, _style2.default['preview_zoom_' + (zoom ? 'out' : 'in')]),
            style: { width: zoom ? preImgWidth : Math.floor(preImgWidth * 0.45) },
            onClick: handleZoom
          },
          Array.isArray(images) && images.map(function (v, idx) {
            return _react2.default.createElement(
              'li',
              { key: idx, className: _style2.default.preview_items },
              _react2.default.createElement('img', { src: v, className: _style2.default.preview_imgs })
            );
          })
        )
      ) : null
    )
  );
};

exports.default = previewImage;
module.exports = exports['default'];
