// @jxkang/utils/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Common = exports.paramType = exports.$getScript = exports.$get = exports.$post = exports.$ajax = undefined;

var _request = require('./request');

var _request2 = _interopRequireDefault(_request);

var _paramType = require('./param-type');

var _paramType2 = _interopRequireDefault(_paramType);

var _common = require('./common');

var Common = _interopRequireWildcard(_common);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

window.jxk_utils_version = '0.0.16';

exports.default = {
  $ajax: _request2.default,
  $post: _request.$post,
  $get: _request.$get,
  $getScript: _request.$getScript,
  paramType: _paramType2.default,
  Common: Common
};
exports.$ajax = _request2.default;
exports.$post = _request.$post;
exports.$get = _request.$get;
exports.$getScript = _request.$getScript;
exports.paramType = _paramType2.default;
exports.Common = Common;
