// @jxkang/web-cmpt/lib/run-second/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _style = require('./style');

var _style2 = _interopRequireDefault(_style);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; } /**
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * @Author: 谭生虎 TanShenghu tanshenghu@163.com
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * @Update: 2019-12-26
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * @Description: 商城秒杀活动
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                */


var noop = function noop() {};

var RunSecond = function (_React$Component) {
  _inherits(RunSecond, _React$Component);

  function RunSecond(props) {
    _classCallCheck(this, RunSecond);

    var _this = _possibleConstructorReturn(this, (RunSecond.__proto__ || Object.getPrototypeOf(RunSecond)).call(this, props));

    _this.runTime = function () {
      var that = _this;
      var _this$props$onFinish = _this.props.onFinish,
          onFinish = _this$props$onFinish === undefined ? noop : _this$props$onFinish;
      var deadline = _this.state.deadline;


      clearTimeout(that.timer);
      var loopTime = function loopTime() {
        that.timer = setTimeout(function () {
          var curNow = deadline - Date.now();
          var state = dateTrans({ deadline: deadline, curNow: curNow });
          that.setState({ surplus: state });
          // ===
          if (state.every(function (item) {
            return item === '00';
          })) {
            clearTimeout(that.timer);
            onFinish('just');
          } else if (curNow > 0) {
            loopTime();
          }
        }, 1000);
      };

      if (deadline && Date.now() <= deadline) {
        loopTime();
      } else {
        onFinish('past');
      }
    };

    _this.state = {
      surplus: ['00', '00', '00', '00'],
      deadline: props.deadline
    };
    _this.timer = null;
    return _this;
  }

  _createClass(RunSecond, [{
    key: 'UNSAFE_componentWillReceiveProps',
    value: function UNSAFE_componentWillReceiveProps(np) {
      if (np.deadline) {
        this.state.deadline = np.deadline;
        this.setState({ deadline: np.deadline });
        this.runTime();
      }
    }
  }, {
    key: 'componentDidMount',
    value: function componentDidMount() {
      this.runTime();
    }
  }, {
    key: 'render',
    value: function render() {
      var _props = this.props,
          className = _props.className,
          dayClassName = _props.dayClassName,
          hourClassName = _props.hourClassName,
          minuteClassName = _props.minuteClassName,
          secondClassName = _props.secondClassName,
          itemsClassName = _props.itemsClassName;
      var surplus = this.state.surplus;


      var cmap = {
        d: dayClassName,
        h: hourClassName,
        m: minuteClassName,
        s: secondClassName
      };
      var renderItems = function renderItems(v, t) {
        return v.split('').map(function (vv, i) {
          return _react2.default.createElement(
            'var',
            { key: i, className: (0, _classnames2.default)(cmap[t] || itemsClassName) },
            vv
          );
        });
      };

      return _react2.default.createElement(
        'div',
        { className: (0, _classnames2.default)(_style2.default.run_second_box, className) },
        renderItems(surplus[0], 'd'),
        _react2.default.createElement(
          'label',
          null,
          '\u5929'
        ),
        renderItems(surplus[1], 'h'),
        _react2.default.createElement(
          'label',
          null,
          '\u65F6'
        ),
        renderItems(surplus[2], 'm'),
        _react2.default.createElement(
          'label',
          null,
          '\u5206'
        ),
        renderItems(surplus[3], 's'),
        _react2.default.createElement(
          'label',
          null,
          '\u79D2'
        )
      );
    }
  }]);

  return RunSecond;
}(_react2.default.Component);

RunSecond.getDate = function (deadline) {
  return dateTrans({ deadline: deadline });
};

function dateTrans() {
  var _ref = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
      deadline = _ref.deadline,
      curNow = _ref.curNow;

  var tm = {
    day: 86400000,
    hour: 3600000,
    minute: 60000,
    second: 1000
  };
  var fill0 = function fill0(v) {
    return ('' + v).length === 1 ? '0' + v : '' + v;
  };
  curNow = typeof curNow === 'undefined' ? deadline - Date.now() : curNow;
  if (curNow < 0) {
    return ['00', '00', '00', '00'];
  }
  // 天数
  var d = fill0(Math.floor(curNow / tm.day));
  curNow %= tm.day;
  // 时
  var h = fill0(Math.floor(curNow / tm.hour));
  curNow %= tm.hour;
  // 分
  var m = fill0(Math.floor(curNow / tm.minute));
  curNow %= tm.minute;
  // 秒
  var s = fill0(Math.floor(curNow / tm.second));
  return ['' + d, '' + h, '' + m, '' + s];
}

exports.default = RunSecond;
module.exports = exports['default'];
