// @jxkang/web-cmpt/lib/pay/pay.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _alert = require('antd/lib/alert');

var _alert2 = _interopRequireDefault(_alert);

var _modal = require('antd/lib/modal');

var _modal2 = _interopRequireDefault(_modal);

var _message2 = require('antd/lib/message');

var _message3 = _interopRequireDefault(_message2);

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _slicedToArray = function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; }(); /**
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          * @Author: 谭生虎 TanShenghu tanshenghu@163.com
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          * @Description: 订单列表 去支付
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          */


require('antd/lib/alert/style');

require('antd/lib/modal/style');

require('antd/lib/message/style');

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _utils = require('@jxkang/utils');

var _formControl = require('../form-control');

var _formControl2 = _interopRequireDefault(_formControl);

var _style = require('./style');

var _style2 = _interopRequireDefault(_style);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var G = window;
var noop = function noop() {};
var Pay = function Pay(_ref) {
  var okText = _ref.okText,
      _ref$type = _ref.type,
      type = _ref$type === undefined ? '1' : _ref$type,
      _ref$amount = _ref.amount,
      amount = _ref$amount === undefined ? 1 : _ref$amount,
      visible = _ref.visible,
      _ref$onOk = _ref.onOk,
      onOk = _ref$onOk === undefined ? noop : _ref$onOk,
      _ref$onCancel = _ref.onCancel,
      onCancel = _ref$onCancel === undefined ? noop : _ref$onCancel,
      url = _ref.url,
      _ref$setParams = _ref.setParams,
      setParams = _ref$setParams === undefined ? function (v) {
    return v;
  } : _ref$setParams,
      _ref$params = _ref.params,
      params = _ref$params === undefined ? {} : _ref$params,
      _ref$directChange = _ref.directChange,
      directChange = _ref$directChange === undefined ? noop : _ref$directChange,
      _ref$unionPage = _ref.unionPage,
      unionPage = _ref$unionPage === undefined ? 'https://static.jingxiaokang.com/assets/html/unionPay.html' : _ref$unionPage;

  okText = okText || (type === '1' ? '确定充值' : type === '2' ? '确定支付' : '充值并支付');
  /**
   * 支付方式
   */
  var payDataSource = [{ label: '微信支付', value: '01' }, { label: '支付宝支付', value: '02' }, { label: '银联支付', value: 'S20' }];

  var defaultAmount = { paymentAmt: amount, payType: null };
  if (type === '3') {
    defaultAmount.also = type === '3';
  }

  var _useState = (0, _react.useState)(defaultAmount),
      _useState2 = _slicedToArray(_useState, 2),
      amountValue = _useState2[0],
      setAmount = _useState2[1];

  var _useState3 = (0, _react.useState)({ visible: false, payUrl: '' }),
      _useState4 = _slicedToArray(_useState3, 2),
      bankData = _useState4[0],
      setBankData = _useState4[1];

  var _useState5 = (0, _react.useState)(unionPage),
      _useState6 = _slicedToArray(_useState5, 2),
      unionPageUrl = _useState6[0],
      setUnionPageUrl = _useState6[1];

  var _useState7 = (0, _react.useState)({}),
      _useState8 = _slicedToArray(_useState7, 2),
      unionPayStateParams = _useState8[0],
      setUnionPayStateParams = _useState8[1];

  var isIE = !/ MSIE /.test(G.navigator.userAgent);
  var needIE = amountValue.payType === 'S20' && isIE; // 需要用IE打开
  var userToken = _utils.Common.getCookie(G.fetchTokenName || 'token');

  var onSubmit = function onSubmit() {
    var reqModel = setParams(_extends({}, amountValue, params));
    if (!reqModel.payType) {
      return _message3.default.warn('请选择支付方式');
    }
    if (!reqModel.paymentAmt) {
      return _message3.default.warn('请输入金额');
    }
    url(reqModel).then(function (resModel) {
      onOk(amountValue, resModel);

      var unionPayParams = _utils.Common.clone(resModel);
      delete unionPayParams.url;

      var _ = function _(vaule) {
        return vaule.replace(/[A-Z]/g, function (a) {
          return '_' + a.toLowerCase();
        });
      };

      // 支付字段处理
      var postUnionPayParams = {};
      Object.keys(unionPayParams).forEach(function (key) {
        postUnionPayParams[_(key)] = unionPayParams[key];
      });

      setUnionPayStateParams(_extends({}, postUnionPayParams));

      /**
       * 支付方式为 银联时 判断是否为IE浏览器
       */
      if (needIE) {
        bankData.payUrl = resModel.url;
        bankData.visible = true;

        unionPage += '?t=' + userToken + '&' + _utils.Common.queryString(reqModel);

        setUnionPageUrl(unionPage);
        setBankData(_extends({}, bankData));
        return false;
      }

      if (resModel && resModel.url) {
        _modal2.default.confirm({
          title: '系统温馨提示',
          content: '您确定要立即支付',
          okText: '立即支付',
          cancelText: '取消',
          onOk: function onOk() {
            _utils.Common.winOpen({
              url: resModel.url,
              type: 'post',
              params: postUnionPayParams
            });
          }
        });
      }
    });
  };

  (0, _react.useEffect)(function () {
    amountValue.paymentAmt = amount;
    setAmount(_extends({}, amountValue));
  }, [amount]);

  // 复制 url
  var onCopyPayUrl = function onCopyPayUrl() {
    _utils.Common.copyText(amountValue.payType === 'S20' ? unionPageUrl : bankData.payUrl);
    _message3.default.success('复制成功');
  };

  // 直接打开支付url
  var onOpenPayUrl = function onOpenPayUrl() {
    _utils.Common.winOpen({
      url: bankData.payUrl,
      type: 'post',
      params: unionPayStateParams
    });
  };

  return _react2.default.createElement(
    _react.Fragment,
    null,
    _react2.default.createElement(
      _modal2.default,
      {
        visible: visible,
        title: type === '1' ? '充值' : '支付',
        okText: okText,
        cancelText: '\u53D6\u6D88',
        okType: 'danger',
        width: 420,
        onOk: onSubmit,
        onCancel: onCancel
      },
      _react2.default.createElement(
        'p',
        { className: _style2.default.title },
        '\u9009\u62E9\u652F\u4ED8\u65B9\u5F0F\uFF1A'
      ),
      _react2.default.createElement(_formControl2.default, {
        type: 'select',
        dataSource: payDataSource,
        placeholder: '\u8BF7\u9009\u62E9',
        width: 372,
        onChange: function onChange(v) {
          amountValue.payType = v;
          setAmount(_extends({}, amountValue));
        },
        className: _style2.default.input
      }),
      _react2.default.createElement(
        'p',
        { className: _style2.default.title },
        type === '1' ? '输入充值金额：' : '输入支付金额：'
      ),
      _react2.default.createElement(_formControl2.default, {
        type: 'number'
        // width={372}
        , style: { width: 372 }
        // prefix={<span>&yen;</span>}
        , placeholder: '\u8BF7\u8F93\u5165\u91D1\u989D',
        className: _style2.default.input,
        trim: true,
        value: amountValue.paymentAmt,
        formatter: function formatter(value) {
          return ('\xA5 ' + value).replace(/\B(?=(\d{3})+(?!\d))/g, ',');
        },
        parser: function parser(value) {
          return value.replace(/¥\s?|(,*)/g, '');
        },
        min: amount,
        onChange: function onChange(v) {
          v = Number(v);
          if (v >= amount) {
            amountValue.paymentAmt = v;
            setAmount(_extends({}, amountValue));
          }
        }
      }),
      type === '3' ? _react2.default.createElement(
        'div',
        null,
        _react2.default.createElement(
          _formControl2.default,
          { type: 'checkbox',
            defaultChecked: true,
            onChange: function onChange(v) {
              amountValue.also = v;
              directChange(v, amountValue);
              setAmount(_extends({}, amountValue));
            }
          },
          '\u76F4\u63A5\u652F\u4ED8'
        )
      ) : null
    ),
    _react2.default.createElement(
      _modal2.default,
      {
        visible: bankData.visible,
        className: _style2.default.bank_pay_pop,
        footer: null,
        onCancel: function onCancel() {
          bankData.visible = false;
          setBankData(_extends({}, bankData));
        }
      },
      _react2.default.createElement(_alert2.default, {
        message: '\u7CFB\u7EDF\u6E29\u99A8\u63D0\u793A',
        description: _react2.default.createElement(
          _react.Fragment,
          null,
          _react2.default.createElement(
            'div',
            { style: { marginBottom: 10 } },
            '\u94F6\u8054\u652F\u4ED8,\u53EF\u80FD\u4EC5\u652F\u6301IE\u6D4F\u89C8\u5668,\u8BF7\u590D\u5236\u4E0B\u9762\u7684\u5730\u5740\u5728IE\u6D4F\u89C8\u5668\u4E2D\u8BBF\u95EE'
          ),
          _react2.default.createElement(
            'div',
            { className: _style2.default.payUrl },
            bankData.payUrl
          ),
          _react2.default.createElement(
            'span',
            {
              className: _style2.default.copy_btn,
              onClick: onOpenPayUrl
            },
            '\u6253\u5F00'
          ),
          _react2.default.createElement(
            'span',
            {
              className: _style2.default.copy_btn,
              onClick: onCopyPayUrl
            },
            '\u590D\u5236'
          )
        ),
        type: 'warning',
        showIcon: true
      })
    )
  );
};

exports.default = Pay;
module.exports = exports['default'];
