// @jxkang/web-cmpt/lib/form-control/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _datePicker = require('antd/lib/date-picker');

var _datePicker2 = _interopRequireDefault(_datePicker);

var _inputNumber = require('antd/lib/input-number');

var _inputNumber2 = _interopRequireDefault(_inputNumber);

var _switch = require('antd/lib/switch');

var _switch2 = _interopRequireDefault(_switch);

var _upload = require('antd/lib/upload');

var _upload2 = _interopRequireDefault(_upload);

var _message2 = require('antd/lib/message');

var _message3 = _interopRequireDefault(_message2);

var _checkbox = require('antd/lib/checkbox');

var _checkbox2 = _interopRequireDefault(_checkbox);

var _radio = require('antd/lib/radio');

var _radio2 = _interopRequireDefault(_radio);

var _select = require('antd/lib/select');

var _select2 = _interopRequireDefault(_select);

var _input = require('antd/lib/input');

var _input2 = _interopRequireDefault(_input);

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; /**
                                                                                                                                                                                                                                                                   * @Author: 谭生虎 TanShenghu tanshenghu@163.com
                                                                                                                                                                                                                                                                   * @Update: 2019-12-14
                                                                                                                                                                                                                                                                   * @Description: 表单元素组件
                                                                                                                                                                                                                                                                   */


require('antd/lib/date-picker/style');

require('antd/lib/input-number/style');

require('antd/lib/switch/style');

require('antd/lib/upload/style');

require('antd/lib/message/style');

require('antd/lib/checkbox/style');

require('antd/lib/radio/style');

require('antd/lib/select/style');

require('antd/lib/input/style');

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _FormBinderWrapper = require('@icedesign/form-binder/lib/FormBinderWrapper');

var _FormBinderWrapper2 = _interopRequireDefault(_FormBinderWrapper);

var _FormBinder = require('@icedesign/form-binder/lib/FormBinder');

var _FormBinder2 = _interopRequireDefault(_FormBinder);

var _FormError = require('@icedesign/form-binder/lib/FormError');

var _FormError2 = _interopRequireDefault(_FormError);

var _utils = require('@jxkang/utils');

var _lang = require('../lang');

var _lang2 = _interopRequireDefault(_lang);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

window.webConfig = window.webConfig || {};
var noop = function noop() {};

/**
 * input
 */
var CustomInput = function CustomInput(props) {
  var controlConf = props.controlConf,
      _props$onChange = props.onChange,
      _onChange = _props$onChange === undefined ? noop : _props$onChange;

  var style = Object.assign({}, { width: props.width }, props.style);

  return _react2.default.createElement(_input2.default, _extends({}, props, controlConf, {
    style: style,
    onChange: function onChange(e) {
      var value = props.trim ? e.target.value.trim() : e.target.value;
      _onChange(value);
    }
  }));
};

/**
 * textarea
 */
var CustomTextarea = function CustomTextarea(props) {
  var controlConf = props.controlConf,
      _props$onChange2 = props.onChange,
      _onChange2 = _props$onChange2 === undefined ? noop : _props$onChange2;

  var style = Object.assign({}, { width: props.width, height: props.height }, props.style);

  return _react2.default.createElement(_input2.default.TextArea, _extends({}, props, controlConf, {
    style: style,
    onChange: function onChange(e) {
      var value = props.trim ? e.target.value.trim() : e.target.value;
      _onChange2(value);
    }
  }));
};

/**
 * select
 */
var CustomSelect = function CustomSelect(props) {
  var controlConf = props.controlConf,
      _props$onChange3 = props.onChange,
      _onChange3 = _props$onChange3 === undefined ? noop : _props$onChange3;

  var style = Object.assign({}, { width: props.width }, props.style);

  return Array.isArray(props.dataSource) ? _react2.default.createElement(
    _select2.default,
    _extends({}, props, { style: style, onChange: function onChange() {
        return _onChange3.apply(undefined, arguments);
      } }, controlConf),
    props.dataSource.map(function (item, idx) {
      return _react2.default.createElement(
        _select2.default.Option,
        { key: idx, value: item.value, disabled: item.disabled },
        item.label
      );
    })
  ) : _react2.default.createElement(_select2.default, _extends({}, props, { style: style }, controlConf));
};

/**
 * radio 组合
 */
var CustomRadio = function CustomRadio(props) {
  var controlConf = props.controlConf,
      _props$onChange4 = props.onChange,
      _onChange4 = _props$onChange4 === undefined ? noop : _props$onChange4,
      value = props.value,
      val = props.val,
      params = _objectWithoutProperties(props, ['controlConf', 'onChange', 'value', 'val']);

  var ivalue = value || val;
  return _react2.default.createElement(_radio2.default, _extends({}, params, { onChange: function onChange(e) {
      return _onChange4(typeof ivalue === 'undefined' ? e.target.checked : e.target.checked ? ivalue : undefined, e);
    } }, controlConf));
};

CustomRadio.Group = function (props) {
  var controlConf = props.controlConf,
      _props$onChange5 = props.onChange,
      _onChange5 = _props$onChange5 === undefined ? noop : _props$onChange5;

  return Array.isArray(props.dataSource) ? _react2.default.createElement(
    _radio2.default.Group,
    _extends({}, props, { onChange: function onChange(e) {
        return _onChange5(e.target.value, e);
      } }, controlConf),
    props.dataSource.map(function (item, idx) {
      return _react2.default.createElement(
        _radio2.default,
        { key: idx, value: item.value },
        item.label
      );
    })
  ) : _react2.default.createElement(_radio2.default.Group, _extends({}, props, { onChange: function onChange(e) {
      return _onChange5(e.target.value, e);
    } }));
};

/**
 * checkbox 组合
 */
var CustomCheckbox = function CustomCheckbox(props) {
  var controlConf = props.controlConf,
      _props$onChange6 = props.onChange,
      _onChange6 = _props$onChange6 === undefined ? noop : _props$onChange6,
      value = props.value,
      val = props.val,
      params = _objectWithoutProperties(props, ['controlConf', 'onChange', 'value', 'val']);

  var ivalue = value || val;
  // e.target.value
  return _react2.default.createElement(_checkbox2.default, _extends({}, params, { onChange: function onChange(e) {
      return _onChange6(typeof ivalue === 'undefined' ? e.target.checked : e.target.checked ? ivalue : undefined, e);
    } }, controlConf));
};

CustomCheckbox.Group = function (props) {
  var controlConf = props.controlConf,
      _props$onChange7 = props.onChange,
      _onChange7 = _props$onChange7 === undefined ? noop : _props$onChange7;

  return Array.isArray(props.dataSource) ? _react2.default.createElement(
    _checkbox2.default.Group,
    _extends({}, props, { onChange: function onChange(e) {
        return _onChange7(e.target.value, e);
      } }, controlConf),
    props.dataSource.map(function (item, idx) {
      return _react2.default.createElement(
        _checkbox2.default,
        { key: idx, value: item.value },
        item.label
      );
    })
  ) : _react2.default.createElement(_checkbox2.default.Group, _extends({}, props, { onChange: function onChange(e) {
      return _onChange7(e.target.value, e);
    } }));
};

// upload组件
var CustomUpload = function CustomUpload(_ref) {
  var value = _ref.value,
      controlConf = _ref.controlConf,
      props = _objectWithoutProperties(_ref, ['value', 'controlConf']);

  var maxSize = props.maxSize,
      _props$beforeUpload = props.beforeUpload,
      _beforeUpload = _props$beforeUpload === undefined ? noop : _props$beforeUpload,
      _props$onUpload = props.onUpload,
      onUpload = _props$onUpload === undefined ? noop : _props$onUpload;

  props.headers = _extends({}, window.webConfig.fetchHeaders || { jdxiaokang_client: 'PC' }, {
    token: _utils.Common.getCookie(window.webConfig.fetchTokenName)
  }, props.headers);
  var baseProps = {
    action: _utils.$ajax.getBaseUrl() + '/productservice/img/upload',
    name: 'file',
    accept: '.jpg,.jpeg,.png',
    beforeUpload: function beforeUpload(file, files) {
      if (typeof maxSize === 'number' && file.size > maxSize) {
        var uni = maxSize / 1024 / 1024 > 1 ? maxSize / 1024 / 1024 + 'M' : maxSize / 1024 + 'KB';
        _message3.default.warn('\u4E0A\u4F20\u6587\u4EF6\u5927\u5C0F\u5DF2\u8D85\u8FC7' + uni);
        return false;
      }
      return _beforeUpload(file, files);
    },

    showUploadList: false,
    multiple: false
  };

  var uploadProps = _extends({}, baseProps, controlConf, props);

  var uploadOnChange = function uploadOnChange() {
    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    onUpload.apply(undefined, args);
    if (args[0].file.status === 'done') {
      var response = args[0].file.response;

      var code = '' + response.responseCode;
      if (code === '0') {
        uploadProps.onChange(response.entry.filePath, args[0], args);
      } else {
        _message3.default.error(response.message || '系统繁忙，请稍后再试');
      }
    }
  };

  return _react2.default.createElement(_upload2.default, _extends({}, uploadProps, {
    onChange: function onChange() {
      return uploadOnChange.apply(undefined, arguments);
    }
  }));
};

var FormControl = function FormControl(props) {
  props.type = props.type || 'text';
  if (props.type && ('' + props.type).indexOf('date') === 0) {
    props.locale = props.locale || _lang2.default.DatePicker;
  }

  var type = props.type,
      name = props.name,
      required = props.required,
      verifMessage = props.verifMessage,
      validator = props.validator,
      pattern = props.pattern,
      oType = props.oType,
      _props$errorClass = props.errorClass,
      errorClass = _props$errorClass === undefined ? '' : _props$errorClass,
      _props$control = props.control,
      control = _props$control === undefined ? {} : _props$control,
      controlProps = _objectWithoutProperties(props, ['type', 'name', 'required', 'verifMessage', 'validator', 'pattern', 'oType', 'errorClass', 'control']);

  var Controls = {
    text: CustomInput,
    textarea: CustomTextarea,
    select: CustomSelect,
    radio: CustomRadio,
    'radio-group': CustomRadio.Group,
    checkbox: CustomCheckbox,
    'checkbox-group': CustomCheckbox.Group,
    switch: _switch2.default,
    number: _inputNumber2.default,
    date: _datePicker2.default,
    'date-range': _datePicker2.default.RangePicker,
    file: CustomUpload,
    empty: _react2.default.Fragment
  };

  var Control = Controls[type];

  // rules
  return name ? _react2.default.createElement(
    _react2.default.Fragment,
    null,
    _react2.default.createElement(
      _FormBinder2.default,
      {
        name: name,
        required: required,
        message: verifMessage,
        validator: validator,
        pattern: pattern
      },
      _react2.default.createElement(Control, _extends({}, controlProps, { type: oType }, control, { controlConf: control }))
    ),
    validator || required || pattern ? _react2.default.createElement(
      'div',
      { className: ('formControl-error-tip ' + errorClass).trim() },
      _react2.default.createElement(_FormError2.default, { name: name })
    ) : null
  ) : _react2.default.createElement(Control, _extends({}, controlProps, { type: oType }, control, { controlConf: control }));
};

FormControl.Wrapper = _FormBinderWrapper2.default;
FormControl.FormBinder = _FormBinder2.default;
FormControl.FormError = _FormError2.default;
exports.default = FormControl;
module.exports = exports['default'];
